cmd/kyverno/main.go

```diff
@@ -396,44 +424,83 @@ func main() {
 		os.Exit(1)
 	}
 
-	// Start the components
-	pInformer.Start(stopCh)
-	kubeInformer.Start(stopCh)
-	kubedynamicInformer.Start(stopCh)
+	// wrap all controllers that need leaderelection
+	// start them once by the leader
+	run := func() {
+		go certManager.Run(stopCh)
+		go policyCtrl.Run(2, prgen.ReconcileCh, stopCh)
+		go prgen.Run(1, stopCh)
+		go grc.Run(genWorkers, stopCh)
+		go grcc.Run(1, stopCh)
+	}
+
+	kubeClientLeaderElection, err := utils.NewKubeClient(clientConfig)
+	if err != nil {
+		setupLog.Error(err, "Failed to create kubernetes client")
+		os.Exit(1)
+	}
+
+	// cleanup Kyverno managed resources followed by webhook shutdown
+	// No need to exit here, as server.Stop(ctx) closes the cleanUp
+	// chan, thus the main process exits.
+	stop := func() {
+		c, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		server.Stop(c)
+	}
+
+	le, err := leaderelection.New("kyverno", config.KyvernoNamespace, kubeClientLeaderElection, run, stop, log.Log.WithName("kyverno/LeaderElection"))
+	if err != nil {
+		setupLog.Error(err, "failed to elector leader")
+		os.Exit(1)
+	}
+
+	// init events handlers
+	// start Kyverno controllers
+	go le.Run(ctx)
 
 	go reportReqGen.Run(2, stopCh)
-	go prgen.Run(1, stopCh)
 	go configData.Run(stopCh)
-	go policyCtrl.Run(2, prgen.ReconcileCh, stopCh)
 	go eventGenerator.Run(3, stopCh)
 	go grgen.Run(10, stopCh)
```

Comment:
@realshuting Do we still need this ?
Since it's not doing anything. 
```
// Run starts the generate request spec
func (g *Generator) Run(workers int, stopCh <-chan struct{}) {
	logger := g.log
	defer utilruntime.HandleCrash()

	logger.V(4).Info("starting")
	defer func() {
		logger.V(4).Info("shutting down")
	}()

	if !cache.WaitForCacheSync(stopCh, g.grSynced) {
		logger.Info("failed to sync informer cache")
		return
	}

	<-g.stopCh
}
```

Reply:
No we can remove it.